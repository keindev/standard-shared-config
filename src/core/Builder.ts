import { promises as fs } from 'fs';
import path from 'path';
import { TaskTree } from 'tasktree-cli';
import { Task } from 'tasktree-cli/lib/Task';

import { EntityName, FileType, ISnapshot } from '../types';
import { getFileContent, getFileHash, getFileType } from '../utils/file';
import { merge, parse, stringify } from '../utils/json';
import Config from './Config';

export default class Builder {
  async build(name: string, config: Config): Promise<void> {
    if (!config.isInitialized) await config.init();

    const { dependencies, scripts, outDir } = config;
    const snapshots = await Promise.all(config.paths.map(filePath => this.createSnapshot(filePath, config)));
    const entities = (
      await Promise.all([
        this.writeEntityScript(EntityName.Dependencies, dependencies, outDir),
        this.writeEntityScript(EntityName.Scripts, scripts, outDir),
        this.writeEntityScript(
          EntityName.Snapshots,
          snapshots.sort((a, b) => b.path.split('/').length - a.path.split('/').length || a.path.localeCompare(b.path)),
          outDir
        ),
      ])
    ).filter(Boolean);

    await fs.writeFile(
      `${outDir}/index.js`,
      [
        '/* --------------------------------------------------------------- */',
        '/* This file generated automatically                               */',
        '/* @see https://www.npmjs.com/package/standard-shared-config       */',
        '/* --------------------------------------------------------------- */',
        '',
        '/* eslint-disable */',
        "const sharedConfig = require('standard-shared-config');",
        ...entities.map(entity => `const ${entity} = require('./${entity}');`),
        '',
        `sharedConfig.share({ ${entities.join(', ')} });`,
      ].join('\n')
    );

    await fs.writeFile(`${outDir}/bin/${name}`, ['#!/usr/bin/env node', "require('../index.js');"].join('\n'));
  }

  async process(snapshots: ISnapshot[]): Promise<void> {
    const task = TaskTree.add('Processing config files...');

    await Promise.all(snapshots.map(snapshot => this.shareConfig(snapshot, task)));
    task.complete('Processed configs:');
  }

  private async shareConfig(snapshot: ISnapshot, task: Task): Promise<void> {
    const filePath = path.resolve(process.cwd(), snapshot.path);
    const content = await getFileContent(filePath);

    if (content) {
      const hash = getFileHash(content);

      if (hash !== snapshot.hash) {
        if (snapshot.merge) {
          await fs.writeFile(filePath, this.mergeFilesContent(snapshot, content));
          task.log(`${snapshot.path} merged`);
        } else {
          await fs.writeFile(filePath, snapshot.content);
          task.log(`${snapshot.path} updated`);
        }
      } else {
        task.log(`${snapshot.path} is not changed`);
      }
    } else {
      await fs.mkdir(path.dirname(filePath), { recursive: true });
      await fs.writeFile(filePath, snapshot.content);

      task.log(`${snapshot.path} created`);
    }
  }

  private mergeFilesContent({ type, merge: rules, content: snapshot }: ISnapshot, content: string): string {
    const excludes = Array.isArray(rules) ? new Set(rules) : undefined;
    let result;

    switch (type) {
      case FileType.GLOB:
        result = [...new Set([...content.split('\n'), ...snapshot.split('\n')].filter(Boolean)).values()]
          .map(row => row.trim())
          .sort()
          .join('\n');
        break;
      case FileType.Text:
        result = [...snapshot.split('\n'), ...content.split('\n')].join('\n');
        break;
      case FileType.JSON:
      case FileType.YAML:
        result = stringify(merge({ left: parse(content, type), right: parse(snapshot, type), excludes }), type);
        break;
      default:
        result = snapshot;
        break;
    }

    return result;
  }

  private async writeEntityScript<T>(name: EntityName, values: T[], outDir: string): Promise<string> {
    const filePath = `${outDir}/${name}.js`;

    if (values.length) {
      await fs.writeFile(
        filePath,
        ['/* eslint-disable */', '/* prettier-ignore */', `module.exports = ${stringify(values)}`].join('\n')
      );
    } else {
      await fs.unlink(filePath);
    }

    return values.length ? name : '';
  }

  private async createSnapshot(filePath: string, config: Config): Promise<ISnapshot> {
    const content = (await getFileContent(filePath)) ?? '';
    const relativePath = path.relative(config.root, filePath);
    const type = getFileType(filePath);
    const rule = config.findRule(relativePath);

    return {
      path: relativePath,
      hash: getFileHash(content),
      merge: [FileType.GLOB, FileType.Text].includes(type) && rule ? true : rule ?? false,
      type,
      content,
    };
  }
}
