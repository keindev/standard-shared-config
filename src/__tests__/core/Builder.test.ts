// @see https://github.com/facebook/jest/issues/9430
import { promises as fs } from 'fs';
import { glob } from 'glob';
import path from 'path';

// eslint-disable-next-line node/no-extraneous-import
import { jest } from '@jest/globals';

import Builder from '../../core/Builder.js';

jest.useFakeTimers();

const CONFIG = `
sharedDir: ".config"
outputDir: "lib"

mergeRules:
  ".vscode/launch.json": [ "configurations" ]
  ".gitignore":

ignorePatterns:
  ".gitignore": [".env", "node_modules/"]

executableFiles: [
  ".husky/commit-msg"
]

scripts:
  "test": "jest"
  "build": "tsc"

dependencies:
  - "@types/jest": "1.x"
  - "ts-jest": "26.x"

package:
  manager: npm
  type: module
  types: './lib/index.d.ts'
  exports:
    ".": "./lib/index.js"
    "lib": "./lib/index.js"
  peerDependencies:
    "react": "18.x"
    "react-dom": "18.x"
`;

jest.spyOn(glob, 'sync').mockImplementation(() => ['.config/config1.json', '.config/test/config2.json']);
jest.spyOn(fs, 'access').mockImplementation(() => Promise.resolve());
jest.spyOn(fs, 'unlink').mockImplementation(() => Promise.resolve());
jest.spyOn(fs, 'mkdir').mockImplementation(() => Promise.resolve(''));
jest.spyOn(fs, 'readFile').mockImplementation(filePath => {
  const basename = path.basename(filePath as string);
  let content = '';

  if (basename === '.sharedconfig.yml') content = CONFIG;
  if (basename === 'package.json') content = JSON.stringify({ name: 'test' });

  return Promise.resolve(content);
});

describe('Builder', () => {
  it('build', async () => {
    const builder = new Builder();
    const files: [string, string][] = [];

    jest.spyOn(fs, 'writeFile').mockImplementation((name, data) => {
      files.push([path.relative(process.cwd(), name.toString()), data.toString()]);

      return Promise.resolve();
    });

    await builder.build('.sharedconfig.yml');

    expect(files).toMatchSnapshot();
  });
});
