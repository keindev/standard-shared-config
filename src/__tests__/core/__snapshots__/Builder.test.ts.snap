// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builder build 1`] = `
[
  [
    "lib/parts/dependencies.js",
    "/* eslint-disable */
/* prettier-ignore */
export default [
  [
    "@types/jest",
    "1.x"
  ],
  [
    "ts-jest",
    "26.x"
  ]
]",
  ],
  [
    "lib/parts/scripts.js",
    "/* eslint-disable */
/* prettier-ignore */
export default [
  [
    "test",
    "jest"
  ],
  [
    "build",
    "tsc"
  ]
]",
  ],
  [
    "lib/parts/snapshots.js",
    "/* eslint-disable */
/* prettier-ignore */
export default [
  {
    "path": "test/config2.json",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "merge": false,
    "executable": false,
    "type": "json",
    "content": ""
  },
  {
    "path": "config1.json",
    "hash": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "merge": false,
    "executable": false,
    "type": "json",
    "content": ""
  },
  {
    "path": ".gitignore",
    "hash": "4526306b7856fe85c9b78e58e5160485994d718cdd48ad32643d220ab5897c72",
    "merge": true,
    "executable": false,
    "type": "glob",
    "content": ".env\\nnode_modules/"
  }
]",
  ],
  [
    "lib/index.js",
    "/* --------------------------------------------------------------- */
/* This file generated automatically                               */
/* @see https://www.npmjs.com/package/standard-shared-config       */
/* --------------------------------------------------------------- */

/* eslint-disable */
import SharedConfig from 'standard-shared-config'
import dependencies from './parts/dependencies.js'
import scripts from './parts/scripts.js'
import snapshots from './parts/snapshots.js'

await new SharedConfig().share(".config", { dependencies, scripts, snapshots, package: {
  "manager": "npm",
  "type": "module",
  "types": "./lib/index.d.ts",
  "exports": {
    ".": "./lib/index.js",
    "lib": "./lib/index.js"
  }
} });",
  ],
  [
    "bin/test.js",
    "#!/usr/bin/env node
import '../lib/index.js';",
  ],
  [
    "package.json",
    "{
  "name": "test",
  "type": "module",
  "exports": "./lib/index.js",
  "bin": {
    "test": "bin/test.js"
  }
}
",
  ],
]
`;
